### PR TITLE
Added bitstream validation code for HEVC parser

### PR DESCRIPTION
**Pull request by rrawther**
Added validation code for fixing parser crashes with edge case data
Need full testing with conformance before merging

### CODE REVIEW

File: src/parser/roc_video_parser.h [hunk 1]
```diff
@@ -80,6 +80,21 @@ typedef struct {
 #define INIT_SEI_PAYLOAD_BUF_SIZE 1024 * 1024  // initial SEI payload buffer size, 1 MB
 #define DECODE_BUF_POOL_EXTENSION 2
 
+#define CHECK_ALLOWED_RANGE(val, min, max) { \
+    if (val < min || val > max) { \
+        ERR ("value not in range: " + TOSTR(val) + "allowed<min,max>: " + TOSTR(min) + " " + TOSTR(max));\
+        goto error; \
```
Comment (AryanSalmanpour) [inline src/parser/roc_video_parser.h]: As discussed, please replace the `goto error; ` with `return PARSER_OUT_OF_RANGE;`

File: src/parser/roc_video_parser.h [hunk 2]
```diff
@@ -80,6 +80,21 @@ typedef struct {
...
+    } \
+}
+
+#define CHECK_ALLOWED_MAX(val, max) { \
+    if (val > max) { \
+        ERR ("value greater than maximum allowed value: " + TOSTR(val) + " max: " + TOSTR(max));\
+        goto error; \
```
Comment (AryanSalmanpour) [inline src/parser/roc_video_parser.h]: As discussed, please replace the `goto error;`  with `return PARSER_OUT_OF_RANGE;`

File: src/parser/hevc_parser.cpp [hunk 3]
```diff
@@ -1853,6 +1870,8 @@ ParserResult HevcVideoParser::ParseSliceHeader(uint8_t *nalu, size_t size, HevcS
 #endif // DBGINFO
 
     return PARSER_OK;
+error:
+    return PARSER_WRONG_STATE;
```
Comment (AryanSalmanpour) [inline src/parser/hevc_parser.cpp]: As discussed, please remove these two lines.

**Review by AryanSalmanpour — CHANGES_REQUESTED** [review]

File: src/parser/hevc_parser.cpp [hunk 4]
```diff
@@ -1542,6 +1543,7 @@ ParserResult HevcVideoParser::ParseSliceHeader(uint8_t *nalu, size_t size, HevcS
 
     // Set active VPS, SPS and PPS for the current slice
     m_active_pps_id_ = Parser::ExpGolomb::ReadUe(nalu, offset);
+    CHECK_ALLOWED_MAX(m_active_pps_id_, (MAX_SPS_COUNT - 1));
```
Comment (jeffqjiangNew) [inline src/parser/hevc_parser.cpp]: Should check against MAX_PPS_COUNT, not MAX_SPS_COUNT.

**Review by jeffqjiangNew — CHANGES_REQUESTED** [review]

File: src/parser/hevc_parser.cpp (hunk 4 above)
Comment (rrawther) [reply src/parser/hevc_parser.cpp]: I thought sps_id ranges from 0 - (count-1)

Comment (jeffqjiangNew) [reply src/parser/hevc_parser.cpp]: It is PPS id, not SPS id, that the slice header is referring to: slice_pic_parameter_set_id.

File: src/parser/hevc_parser.cpp [hunk 5]
```diff
@@ -1772,19 +1783,23 @@ ParserResult HevcVideoParser::ParseSliceHeader(uint8_t *nalu, size_t size, HevcS
                 }
                 if ((p_slice_header->collocated_from_l0_flag && p_slice_header->num_ref_idx_l0_active_minus1 > 0) || (!p_slice_header->collocated_from_l0_flag && p_slice_header->num_ref_idx_l1_active_minus1 > 0)) {
                     p_slice_header->collocated_ref_idx = Parser::ExpGolomb::ReadUe(nalu, offset);
+                    CHECK_ALLOWED_MAX(p_slice_header->collocated_ref_idx, p_slice_header->num_ref_idx_l0_active_minus1);
```
Comment (jeffqjiangNew) [inline src/parser/hevc_parser.cpp]: Should check against num_ref_idx_l0_active_minus1 or num_ref_idx_l1_active_minus1 according to slice type. As per  spec:
When slice_type is equal to P or when slice_type is equal to B and collocated_from_l0_flag is equal to 1,
collocated_ref_idx refers to a picture in list 0, and the value of collocated_ref_idx shall be in the range of 0 to
num_ref_idx_l0_active_minus1, inclusive.
When slice_type is equal to B and collocated_from_l0_flag is equal to 0, collocated_ref_idx refers to a picture in list 1
and the value of collocated_ref_idx shall be in the range of 0 to num_ref_idx_l1_active_minus1, inclusive.

**Review by jeffqjiangNew — CHANGES_REQUESTED** [review]

File: src/parser/hevc_parser.cpp (hunk 5 above)
Comment (rrawther) [reply src/parser/hevc_parser.cpp]: Yes. We need to definitely add more error checks. I just added some trivial ones

File: src/parser/hevc_parser.cpp [hunk 6]
```diff
@@ -1772,19 +1784,23 @@ ParserResult HevcVideoParser::ParseSliceHeader(uint8_t *nalu, size_t size, HevcS
...
                 }
             }
 
             if ((pps_ptr->weighted_pred_flag && p_slice_header->slice_type == HEVC_SLICE_TYPE_P) || (pps_ptr->weighted_bipred_flag && p_slice_header->slice_type == HEVC_SLICE_TYPE_B)) {
                 ParsePredWeightTable(p_slice_header, chroma_array_type, nalu, offset);
             }
             p_slice_header->five_minus_max_num_merge_cand = Parser::ExpGolomb::ReadUe(nalu, offset);
+            //CHECK_ALLOWED_MAX(p_slice_header->five_minus_max_num_merge_cand, 4);
```
Comment (jeffqjiangNew) [inline src/parser/hevc_parser.cpp:1799]: This is correct. We can uncomment it.

File: src/parser/hevc_parser.cpp (hunk 5 above)
Comment (jeffqjiangNew) [reply src/parser/hevc_parser.cpp]: Here is the code to replace the above line:

if (p_slice_header->slice_type == HEVC_SLICE_TYPE_P || (p_slice_header->slice_type == HEVC_SLICE_TYPE_B && p_slice_header->collocated_from_l0_flag == 1)) {
                        CHECK_ALLOWED_MAX(p_slice_header->collocated_ref_idx, p_slice_header->num_ref_idx_l0_active_minus1);
                    } else if (p_slice_header->slice_type == HEVC_SLICE_TYPE_B && p_slice_header->collocated_from_l0_flag == 0) {
                        CHECK_ALLOWED_MAX(p_slice_header->collocated_ref_idx, p_slice_header->num_ref_idx_l1_active_minus1);
                    }

Comment (jeffqjiangNew) [reply src/parser/hevc_parser.cpp]: With the above two requested changes, all HEVC conformance streams are now passing.

Comment (rrawther) [reply src/parser/hevc_parser.cpp]: replaced it

File: src/parser/hevc_parser.cpp (hunk 6 above)
Comment (rrawther) [reply src/parser/hevc_parser.cpp:1799]: It was causing failures in streams. So I have commented it

**Review by LakshmiKumar23 — APPROVED** [review]
Performance looks consistent

**Review by AryanSalmanpour — APPROVED** [review]

**Review by jeffqjiangNew — APPROVED** [review]

**Pull request merged**